### ISSUE TITLE
Read CSV files created by R (header line is one line shorter)

### ISSUE THREAD

**Issue by ghuls** [pola-rs/polars#1999]
#### Describe your feature request

Read CSV files created by R. R does create annoying CSV files by default, where they don't write the column name for the index column.

Somewhere in https://github.com/pola-rs/polars/blob/master/polars/polars-io/src/csv_core/utils.rs#L159-L190 it should read the line after the header to determine the real number of columns (same number of columns as in header or same number of columns in header +1).

```
❯  head test*.txt
==> test.txt <==
1   6   a
2   7   b
3   8   c
4   9   d
5   10  e

==> test2.txt <==
"V1" "V2" "V3"
"1" 1 6 "a"
"2" 2 7 "b"
"3" 3 8 "c"
"4" 4 9 "d"
"5" 5 10 "e"
```

```R
> read.table('test.txt')
  V1 V2 V3
1  1  6  a
2  2  7  b
3  3  8  c
4  4  9  d
5  5 10  e
> df <- read.table('test.txt')
> write.table(df, 'test2.txt')
```

**Comment by ritchie46** [comment]
But then again, if we would read that I would argue that it should read the header as `V1 V2 V3, None`.

**Comment by ghuls** [comment]
`None, V1, V2, V3` would make more sense for those files.

**Comment by ghuls** [comment]
A lot of "CSV" files uploaded to [GEO](https://www.ncbi.nlm.nih.gov/geo/) are files written in R with this format.

e.g files inside `GSE120000_RAW.tar`: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?&acc=GSE120000

**Comment by ghuls** [comment]
@ritchie46 
Now that having CSV files with an extra field in the header, supporting the other way around (one header less than columns as first column is index) would make sense too. First column could be called "row_nr" or "row_index" (not sure if it will be always a number or not).
https://github.com/pola-rs/polars/pull/5210/commits/0d7bf8929edded31d96ae859af1cfb6cc6549e2d

**Comment by Wainberg** [comment]
Here's a minimal example:

```
import polars as pl
from io import StringIO
pl.read_csv(StringIO('a\n1,2'))
```

which currently gives:

```
polars.exceptions.ComputeError: found more fields than defined in 'Schema'

Consider setting 'truncate_ragged_lines=True'.
```

**Comment by Wainberg** [comment]
This is one of those things that would be good to support if you want bioinformatics folks to migrate en masse to polars (which they should!). 

CSV files from R are ubiquitous, and so are files from GEO (which is the repository for pretty much all RNA sequencing data in existence) as @ghuls points out. Pandas supports them too. 

They're pretty much one of the standard formats for CSV files. I wouldn't even argue they're malformed, considering that one of the two main data science programming languages uses them as their default format!

If there's interest, I'd be very glad to write a pull request for this.

**Comment by stinodego** [comment]
I think this would be good to support. @Wainberg feel free to set up a PR for this!

**Comment by ritchie46** [comment]
I don't think it is a proper csv file if it misses a column. I am open to a keyword argument that allows extending the schema. But  I still feel strongly that this should error by default.

**Comment by Wainberg** [comment]
Besides R and pandas, this CSV format is also supported by dask:

```python
>>> import dask.dataframe as dd
>>> dd.read_csv('foo.csv').compute()
   a
1  2
```

In practice, there's no ambiguity: the column headers correspond to the second to Nth columns, and the first column header is blank. This is a convention of CSV files. Polars should respect it.

**Comment by ritchie46** [comment]
> Polars should respect it.

We follow the rfc 4180 
https://datatracker.ietf.org/doc/html/rfc4180

What we should en shouldn't isn't defined by what pandas does. Dask follows pandas in this regard. But R and pandas extended a file format with an index.

As I said, I am open to allowing a keyword argument that defines how many fields must be prefixed. I just don't think that should be default. Once we have that prefix we can think of support flavors.

**Comment by Wainberg** [comment]
The [CSV](https://en.wikipedia.org/wiki/Comma-separated_values) Wikipedia page notes that:

> The lack of adherence to the CSV standard RFC 4180 necessitates the support for a variety of CSV formats in data input software

Polars should absolutely be strictly compliant with RFC 4180 when writing CSVs (at least with default arguments), but it needs to be able to read non-compliant CSVs somehow. I think we agree on that point.

If enabling support is as simple as adding `flavor='R'` or something, AND that's auto-detected with an error message telling the user to add the flag for only the files that need it, that could work.

But why make it harder for the user if you can just auto-detect it anyway? Strict compliance with a standard many people don't comply with doesn't seem like enough of a reason. "What pandas and R do" is more of a de facto standard than RFC 4180.

**Comment by ritchie46** [comment]
Because I think a missing field in the header is ambiguous and might be a malformed csv. Any of the fields might be missing. I want to be informed of that when I read a file.

**Comment by Wainberg** [comment]
How about reading the file under the assumption that it's an R-style CSV, but give a warning that informs the user that there's one fewer column header than the number of columns, and that they can silence the warning with a boolean flag?

**Comment by stinodego** [comment]
I think the decision is clear, no need to keep dwelling on this. If the header column has a missing field, it's an error by default.

The additional argument can be either a boolean flag or perhaps an optional string indicating the name of the missing index column (default None = error). We can bikeshed the name of the parameter when there's a PR.

**Comment by Wainberg** [comment]
I like the idea of having an `index_column_name` parameter (or whatever you'd like the name to be), with an optional string indicating the name of the index column with the missing header. 

Otherwise you'd have to pick a default value, like `''` (a bit ugly) or `'index'` (could clash with another column named `'index'`). Pandas and R don't have this problem because the index column just becomes the index/rownames, so it's easy to refer to it later.

What would the behavior be when `index_column_name` is not None but the CSV isn't missing any column headers - presumably an error?

**Comment by stinodego** [comment]
> What would the behavior be when index_column_name is not None but the CSV isn't missing any column headers - presumably an error?

Yes!

**Comment by Wainberg** [comment]
Ok sounds good. If Ritchie is ok with your formulation, I'll go ahead and implement it.

**Comment by ritchie46** [comment]
> The additional argument can be either a boolean flag or perhaps an optional string indicating the name of the missing index column (default None = error).

How about two new parameters. `prefix_fields: list[str]` and `suffix_fields: list[str]`? That way we can prepend and/or append any field datatype that has missing fields. That should give the reader the possibility to adapt to different flavors.

**Comment by Wainberg** [comment]
That would certainly give more flexibility. I guess it depends on the relative frequency of each type of CSV "in the wild". 

Anecdotally, R-style CSV files are relatively common - maybe 5-10% of all CSVs - and I don't ever remember encountering files with a column missing at the end, or multiple columns missing at either the start or the end. 

Based on that, I'd say having an `index_column_name` parameter would be enough for most use-cases, and for more esoteric CSVs you could specify `has_header=False`, `skip_rows=1`, and `new_columns`.

**Comment by ritchie46** [comment]
We are going down in the discussion again. I want to tailor the flexibility of the reader. Not add index columns.

I think we should add the before mentioned parameters.

**Comment by Wainberg** [comment]
Sure thing, feel free to add them. I'd advocate for `prefix_fields: str | list[str]` and `suffix_fields: str | list[str]` but it's ultimately your decision.

**Comment by ghuls** [comment]
For most of those R generated CSV files on GEO, the index column is actually useful (gene names instead of just a numeric row ID).

**Comment by srajan0149** [comment]
This issue is also resolved.
[row_index_name](https://docs.pola.rs/api/python/stable/reference/api/polars.read_csv.html#:~:text=header%20is%20parsed.-,row_index_name,set%20to%20None%20(default)%2C%20no%20row%20index%20column%20is%20created.,-row_index_offset)
